File: ios/RNKakaoError.swift
import Foundation
import KakaoSDKAuth
import KakaoSDKCommon
import KakaoSDKUser

enum RNKakaoError {
  typealias ParsedError = (code: String, message: String, sdkMessage: String?)

  // 에러 코드
  private enum Code {
    static let accessDenied = "KAKAO_ACCESS_DENIED"
    static let apiError = "KAKAO_API_ERROR"
    static let authError = "KAKAO_AUTH_ERROR"
    static let badParameter = "KAKAO_BAD_PARAMETER"
    static let cancelled = "KAKAO_CANCELLED"
    static let clientError = "KAKAO_CLIENT_ERROR"
    static let defaultError = "KAKAO_ERROR"
    static let forbidden = "KAKAO_FORBIDDEN"
    static let illegalState = "KAKAO_ILLEGAL_STATE"
    static let invalidAppKey = "KAKAO_INVALID_APP_KEY"
    static let invalidBundleId = "KAKAO_INVALID_BUNDLE_ID"
    static let invalidClient = "KAKAO_INVALID_CLIENT"
    static let invalidGrant = "KAKAO_INVALID_GRANT"
    static let invalidRequest = "KAKAO_INVALID_REQUEST"
    static let invalidScope = "KAKAO_INVALID_SCOPE"
    static let invalidUrlScheme = "KAKAO_INVALID_URL_SCHEME"
    static let loginRequired = "KAKAO_LOGIN_REQUIRED"
    static let misconfigured = "KAKAO_MISCONFIGURED"
    static let notSupported = "KAKAO_NOT_SUPPORTED"
    static let profileNotFound = "KAKAO_PROFILE_NOT_FOUND"
    static let rateLimit = "KAKAO_RATE_LIMIT"
    static let serverError = "KAKAO_SERVER_ERROR"
    static let shippingAddressesNotFound = "KAKAO_SHIPPING_ADDRESSES_NOT_FOUND"
    static let tokenExpired = "KAKAO_TOKEN_EXPIRED"
    static let tokenNotFound = "KAKAO_TOKEN_NOT_FOUND"
    static let unknownLogin = "KAKAO_UNKNOWN_LOGIN"
  }

  // 에러 메시지
  private enum Message {
    static let accessDenied = "Kakao login was cancelled or access was denied."
    static let apiError = "A Kakao API error occurred."
    static let authError = "A Kakao authentication error occurred."
    static let badParameter = "Invalid parameter."
    static let cancelled = "Login was cancelled."
    static let clientError = "A Kakao client error occurred."
    static let defaultError = "An error occurred while processing the Kakao request."
    static let forbidden = "Permission is required. Please check the consent items."
    static let illegalState = "Invalid state."
    static let invalidAppKey = "The Kakao native app key is invalid. Please check your Kakao app key configuration."
    static let invalidBundleId = "The iOS bundle ID is invalid. Please check the iOS bundle ID in Kakao Developers and your Xcode settings."
    static let invalidClient = "The Kakao iOS app configuration is invalid. Please check the app key, bundle ID, URL scheme, and platform settings."
    static let invalidGrant = "The authorization information is invalid. Please sign in again."
    static let invalidRequest = "The Kakao login request is invalid. Please check your iOS app configuration and request parameters."
    static let invalidScope = "The requested scope is invalid. Please check the requested consent items and Kakao Developers settings."
    static let invalidUrlScheme = "The iOS URL scheme is invalid. Please check CFBundleURLTypes, KAKAO_APP_SCHEME, and Kakao Developers settings."
    static let loginRequired = "Kakao login authorization is required. Please check app permissions and platform settings."
    static let misconfigured = "The Kakao iOS configuration is invalid. Please check the bundle ID, URL scheme, KAKAO_APP_KEY, and Kakao Developers settings."
    static let notSupported = "This feature is not supported."
    static let profileNotFound = "Kakao profile information could not be found."
    static let rateLimit = "Too many requests. Please try again later."
    static let serverError = "A Kakao server error occurred. Please try again later."
    static let shippingAddressesNotFound = "Kakao shipping address information could not be found."
    static let tokenExpired = "Authentication has expired. Please sign in again."
    static let tokenNotFound = "Sign-in is required."
    static let unknownLogin = "Kakao login failed."
  }

  // 프로필 없음
  static func profileNotFound() -> ParsedError {
    return make(Code.profileNotFound, Message.profileNotFound)
  }

  // 배송지 없음
  static func shippingAddressesNotFound() -> ParsedError {
    return make(Code.shippingAddressesNotFound, Message.shippingAddressesNotFound)
  }

  // 로그인 실패
  static func unknownLogin() -> ParsedError {
    return make(Code.unknownLogin, Message.unknownLogin)
  }

  // SDK 에러 변환
  static func parse(_ error: Error) -> ParsedError {
    let sdkMessage = resolveSdkMessage(error)

    if let sdkError = error as? SdkError {
      switch sdkError {
      case .ClientFailed(let reason, _):
        return resolveClientError(reason, sdkMessage)
      case .ApiFailed(let reason, _):
        return resolveApiError(reason, sdkMessage)
      case .AuthFailed(let reason, _):
        return resolveAuthError(reason, sdkMessage)
      case .AppsFailed(_, _):
        return make(Code.apiError, Message.apiError, sdkMessage)
      }
    }

    return make(Code.defaultError, Message.defaultError, sdkMessage)
  }

  // 클라이언트 에러 변환
  private static func resolveClientError(_ reason: ClientFailureReason, _ sdkMessage: String?) -> ParsedError {
    switch reason {
    case .Cancelled:
      return make(Code.cancelled, Message.cancelled, sdkMessage)
    case .TokenNotFound:
      return make(Code.tokenNotFound, Message.tokenNotFound, sdkMessage)
    case .NotSupported:
      return make(Code.notSupported, Message.notSupported, sdkMessage)
    case .BadParameter:
      return make(Code.badParameter, Message.badParameter, sdkMessage)
    case .IllegalState:
      return make(Code.illegalState, Message.illegalState, sdkMessage)
    case .CastingFailed:
      return make(Code.clientError, Message.clientError, sdkMessage)
    default:
      return make(Code.clientError, Message.clientError, sdkMessage)
    }
  }

  // API 에러 변환
  private static func resolveApiError(_ reason: ApiFailureReason, _ sdkMessage: String?) -> ParsedError {
    let lowercased = sdkMessage?.lowercased() ?? ""

    if lowercased.contains("token") {
      return make(Code.tokenExpired, Message.tokenExpired, sdkMessage)
    }

    if lowercased.contains("permission") || lowercased.contains("scope") || lowercased.contains("forbidden") {
      return make(Code.forbidden, Message.forbidden, sdkMessage)
    }

    if lowercased.contains("too many") || lowercased.contains("rate") || lowercased.contains("limit") {
      return make(Code.rateLimit, Message.rateLimit, sdkMessage)
    }

    if lowercased.contains("server") || lowercased.contains("internal") {
      return make(Code.serverError, Message.serverError, sdkMessage)
    }

    return make(Code.apiError, Message.apiError, sdkMessage)
  }

  // 인증 에러 변환
  private static func resolveAuthError(_ reason: AuthFailureReason, _ sdkMessage: String?) -> ParsedError {
    switch reason {
    case .InvalidClient:
      return resolveAuthConfigurationError(sdkMessage)
    case .InvalidGrant:
      return make(Code.invalidGrant, Message.invalidGrant, sdkMessage)
    case .InvalidRequest:
      return resolveInvalidRequest(sdkMessage)
    case .InvalidScope:
      return make(Code.invalidScope, Message.invalidScope, sdkMessage)
    case .Misconfigured:
      return make(Code.misconfigured, Message.misconfigured, sdkMessage)
    case .Unauthorized:
      return make(Code.loginRequired, Message.loginRequired, sdkMessage)
    case .AccessDenied:
      return make(Code.accessDenied, Message.accessDenied, sdkMessage)
    case .ServerError:
      return make(Code.serverError, Message.serverError, sdkMessage)
    default:
      return make(Code.authError, Message.authError, sdkMessage)
    }
  }

  // 인증 설정 에러 변환
  private static func resolveAuthConfigurationError(_ sdkMessage: String?) -> ParsedError {
    let lowercased = sdkMessage?.lowercased() ?? ""

    if lowercased.contains("app key") || lowercased.contains("client_id") || lowercased.contains("invalid_client") {
      return make(Code.invalidAppKey, Message.invalidAppKey, sdkMessage)
    }

    if lowercased.contains("bundle") {
      return make(Code.invalidBundleId, Message.invalidBundleId, sdkMessage)
    }

    if lowercased.contains("scheme") || lowercased.contains("redirect") {
      return make(Code.invalidUrlScheme, Message.invalidUrlScheme, sdkMessage)
    }

    return make(Code.invalidClient, Message.invalidClient, sdkMessage)
  }

  // 잘못된 요청 에러 변환
  private static func resolveInvalidRequest(_ sdkMessage: String?) -> ParsedError {
    let lowercased = sdkMessage?.lowercased() ?? ""

    if lowercased.contains("scheme") || lowercased.contains("redirect_uri") {
      return make(Code.invalidUrlScheme, Message.invalidUrlScheme, sdkMessage)
    }

    return make(Code.invalidRequest, Message.invalidRequest, sdkMessage)
  }

  // SDK 원문 메시지 추출
  private static func resolveSdkMessage(_ error: Error) -> String? {
    let message = error.localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
    return message.isEmpty ? nil : message
  }

  // 에러 생성
  private static func make(_ code: String, _ message: String, _ sdkMessage: String? = nil) -> ParsedError {
    return (code: code, message: message, sdkMessage: sdkMessage)
  }
}

File: android/src/main/java/kr/packagekr/kakao/signin/RNKakaoError.kt
package kr.packagekr.kakao.signin

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.kakao.sdk.common.model.ApiError
import com.kakao.sdk.common.model.ApiErrorCause
import com.kakao.sdk.common.model.AuthError
import com.kakao.sdk.common.model.AuthErrorCause
import com.kakao.sdk.common.model.ClientError
import com.kakao.sdk.common.model.ClientErrorCause

object RNKakaoError {
    // 변환된 에러
    data class ParsedError(
        val code: String,
        val message: String,
        val sdkMessage: String? = null,
    )

    // 에러 메시지
    private object Message {
        const val accessDenied = "Kakao login was cancelled or access was denied."
        const val activityDoesNotExist = "The current Activity could not be found."
        const val apiError = "A Kakao API error occurred."
        const val authError = "A Kakao authentication error occurred."
        const val badParameter = "Invalid parameter."
        const val cancelled = "Login was cancelled."
        const val clientError = "A Kakao client error occurred."
        const val defaultError = "An error occurred while processing the Kakao request."
        const val forbidden = "Permission is required. Please check the consent items."
        const val illegalState = "Invalid state."
        const val invalidAppKey = "The Kakao native app key is invalid. Please check your Kakao app key configuration."
        const val invalidBundleId = "The Android package name is invalid. Please check the Android package name in Kakao Developers and your app settings."
        const val invalidClient = "The Kakao Android app configuration is invalid. Please check the app key, package name, key hash, URL scheme, and platform settings."
        const val invalidGrant = "The authorization information is invalid. Please sign in again."
        const val invalidRequest = "The Kakao login request is invalid. Please check your Android app configuration and request parameters."
        const val invalidScope = "The requested scope is invalid. Please check the requested consent items and Kakao Developers settings."
        const val invalidUrlScheme = "The Android URL scheme is invalid. Please check AndroidManifest, kakao_custom_scheme, and Kakao Developers settings."
        const val loginRequired = "Kakao login authorization is required. Please check app permissions and platform settings."
        const val misconfigured = "The Kakao Android configuration is invalid. Please check the app key, package name, key hash, URL scheme, and Kakao Developers settings."
        const val notSupported = "This feature is not supported."
        const val profileNotFound = "Kakao profile information could not be found."
        const val rateLimit = "Too many requests. Please try again later."
        const val serverError = "A Kakao server error occurred. Please try again later."
        const val shippingAddressesNotFound = "Kakao shipping address information could not be found."
        const val tokenExpired = "Authentication has expired. Please sign in again."
        const val tokenNotFound = "Sign-in is required."
        const val unknownLogin = "Kakao login failed."
    }

    // 에러 코드
    const val ACTIVITY_DOES_NOT_EXIST = "KAKAO_ACTIVITY_DOES_NOT_EXIST"
    const val ACCESS_DENIED = "KAKAO_ACCESS_DENIED"
    const val API_ERROR = "KAKAO_API_ERROR"
    const val AUTH_ERROR = "KAKAO_AUTH_ERROR"
    const val BAD_PARAMETER = "KAKAO_BAD_PARAMETER"
    const val CANCELLED = "KAKAO_CANCELLED"
    const val CLIENT_ERROR = "KAKAO_CLIENT_ERROR"
    const val ERROR = "KAKAO_ERROR"
    const val FORBIDDEN = "KAKAO_FORBIDDEN"
    const val ILLEGAL_STATE = "KAKAO_ILLEGAL_STATE"
    const val INVALID_APP_KEY = "KAKAO_INVALID_APP_KEY"
    const val INVALID_BUNDLE_ID = "KAKAO_INVALID_BUNDLE_ID"
    const val INVALID_CLIENT = "KAKAO_INVALID_CLIENT"
    const val INVALID_GRANT = "KAKAO_INVALID_GRANT"
    const val INVALID_REQUEST = "KAKAO_INVALID_REQUEST"
    const val INVALID_SCOPE = "KAKAO_INVALID_SCOPE"
    const val INVALID_URL_SCHEME = "KAKAO_INVALID_URL_SCHEME"
    const val LOGIN_REQUIRED = "KAKAO_LOGIN_REQUIRED"
    const val MISCONFIGURED = "KAKAO_MISCONFIGURED"
    const val NOT_SUPPORTED = "KAKAO_NOT_SUPPORTED"
    const val PROFILE_NOT_FOUND = "KAKAO_PROFILE_NOT_FOUND"
    const val RATE_LIMIT = "KAKAO_RATE_LIMIT"
    const val SERVER_ERROR = "KAKAO_SERVER_ERROR"
    const val SHIPPING_ADDRESSES_NOT_FOUND = "KAKAO_SHIPPING_ADDRESSES_NOT_FOUND"
    const val TOKEN_EXPIRED = "KAKAO_TOKEN_EXPIRED"
    const val TOKEN_NOT_FOUND = "KAKAO_TOKEN_NOT_FOUND"
    const val UNKNOWN = "KAKAO_UNKNOWN_ERROR"

    // 에러 응답 변환
    fun reject(promise: Promise, error: Throwable) {
        val parsed = parse(error)
        val userInfo = Arguments.createMap()

        parsed.sdkMessage?.let {
            userInfo.putString("sdkMessage", it)
        }

        promise.reject(parsed.code, parsed.message, error, userInfo)
    }

    // 직접 에러 응답
    fun rejectActivityDoesNotExist(promise: Promise) {
        promise.reject(ACTIVITY_DOES_NOT_EXIST, Message.activityDoesNotExist)
    }

    fun rejectProfileNotFound(promise: Promise) {
        promise.reject(PROFILE_NOT_FOUND, Message.profileNotFound)
    }

    fun rejectShippingAddressesNotFound(promise: Promise) {
        promise.reject(SHIPPING_ADDRESSES_NOT_FOUND, Message.shippingAddressesNotFound)
    }

    fun rejectUnknownLogin(promise: Promise) {
        promise.reject(UNKNOWN, Message.unknownLogin)
    }

    // SDK 에러 변환
    fun parse(error: Throwable): ParsedError {
        return when (error) {
            is ClientError -> resolveClientError(error)
            is ApiError -> resolveApiError(error)
            is AuthError -> resolveAuthError(error)
            else -> make(ERROR, Message.defaultError, sdkMessage(error))
        }
    }

    // 클라이언트 에러
    private fun resolveClientError(error: ClientError): ParsedError {
        val sdkMessage = sdkMessage(error)

        return when (error.reason) {
            ClientErrorCause.Cancelled ->
                make(CANCELLED, Message.cancelled, sdkMessage)
            ClientErrorCause.NotSupported ->
                make(NOT_SUPPORTED, Message.notSupported, sdkMessage)
            ClientErrorCause.BadParameter ->
                make(BAD_PARAMETER, Message.badParameter, sdkMessage)
            ClientErrorCause.TokenNotFound ->
                make(TOKEN_NOT_FOUND, Message.tokenNotFound, sdkMessage)
            ClientErrorCause.IllegalState ->
                make(ILLEGAL_STATE, Message.illegalState, sdkMessage)
            else -> make(CLIENT_ERROR, Message.clientError, sdkMessage)
        }
    }

    // API 에러
    private fun resolveApiError(error: ApiError): ParsedError {
        val sdkMessage = sdkMessage(error)

        return when (error.reason) {
            ApiErrorCause.InvalidToken ->
                make(TOKEN_EXPIRED, Message.tokenExpired, sdkMessage)
            ApiErrorCause.BlockedAccount, ApiErrorCause.BlockedApp, ApiErrorCause.PermissionDenied, ApiErrorCause.InsufficientScope ->
                make(FORBIDDEN, Message.forbidden, sdkMessage)
            ApiErrorCause.ApiLimitExceeded ->
                make(RATE_LIMIT, Message.rateLimit, sdkMessage)
            ApiErrorCause.IllegalParams ->
                make(BAD_PARAMETER, Message.badParameter, sdkMessage)
            ApiErrorCause.UnsupportedApi, ApiErrorCause.DeprecatedApi ->
                make(NOT_SUPPORTED, Message.notSupported, sdkMessage)
            ApiErrorCause.InternalError, ApiErrorCause.ServerTimeOut, ApiErrorCause.UnderMaintenance ->
                make(SERVER_ERROR, Message.serverError, sdkMessage)
            ApiErrorCause.AppDoesNotExist ->
                make(INVALID_APP_KEY, Message.invalidAppKey, sdkMessage)
            else -> make(API_ERROR, Message.apiError, sdkMessage)
        }
    }

    // 인증 에러
    private fun resolveAuthError(error: AuthError): ParsedError {
        val sdkMessage = error.response.errorDescription?.trim()?.takeIf { it.isNotEmpty() } ?: sdkMessage(error)
        val normalizedDescription = sdkMessage?.lowercase().orEmpty()

        return when (error.reason) {
            AuthErrorCause.InvalidRequest -> {
                when {
                    normalizedDescription.contains("package") || normalizedDescription.contains("bundle") ->
                        make(INVALID_BUNDLE_ID, Message.invalidBundleId, sdkMessage)
                    normalizedDescription.contains("scheme") || normalizedDescription.contains("redirect") ->
                        make(INVALID_URL_SCHEME, Message.invalidUrlScheme, sdkMessage)
                    normalizedDescription.contains("client") || normalizedDescription.contains("app key") ->
                        make(INVALID_APP_KEY, Message.invalidAppKey, sdkMessage)
                    else ->
                        make(INVALID_REQUEST, Message.invalidRequest, sdkMessage)
                }
            }
            AuthErrorCause.InvalidClient ->
                make(INVALID_CLIENT, Message.invalidClient, sdkMessage)
            AuthErrorCause.AccessDenied ->
                make(ACCESS_DENIED, Message.accessDenied, sdkMessage)
            AuthErrorCause.InvalidScope ->
                make(INVALID_SCOPE, Message.invalidScope, sdkMessage)
            AuthErrorCause.Misconfigured ->
                make(MISCONFIGURED, Message.misconfigured, sdkMessage)
            AuthErrorCause.Unauthorized, AuthErrorCause.LoginRequired ->
                make(LOGIN_REQUIRED, Message.loginRequired, sdkMessage)
            AuthErrorCause.ConsentRequired, AuthErrorCause.InteractionRequired ->
                make(FORBIDDEN, Message.forbidden, sdkMessage)
            AuthErrorCause.InvalidGrant ->
                make(INVALID_GRANT, Message.invalidGrant, sdkMessage)
            AuthErrorCause.ServerError ->
                make(SERVER_ERROR, Message.serverError, sdkMessage)
            else -> make(AUTH_ERROR, Message.authError, sdkMessage)
        }
    }

    // SDK 원문 메시지 추출
    private fun sdkMessage(error: Throwable): String? {
        return error.message?.trim()?.takeIf { it.isNotEmpty() }
    }

    // 에러 생성
    private fun make(code: String, message: String, sdkMessage: String? = null): ParsedError {
        return ParsedError(code, message, sdkMessage)
    }
}
